fig/backends/aws/__init__.py

```diff
@@ -1,50 +1,153 @@
+    def __init__(self, event):
         self.event = event
 
+    def find_instance(self, instance_id, mac_address):  # pylint: disable=R0201
+        # Instance IDs are unique to the region, not the account, so we have to check them all
+
+        ec2_client = boto3.client("ec2")
+        regions = [region["RegionName"] for region in ec2_client.describe_regions()["Regions"]]
+        for region in regions:
+            ec2 = boto3.resource("ec2", region_name=region)
+            try:
+                ec2instance = ec2.Instance(instance_id)
+                found = False
+                # Confirm the mac address matches
+                for iface in ec2instance.network_interfaces:
+                    det_mac = mac_address.lower().replace(":", "").replace("-", "")
+                    ins_mac = iface.mac_address.lower().replace(":", "").replace("-", "")
+                    if det_mac == ins_mac:
+                        found = True
+                if found:  # pylint: disable=R1723
+                    break
+                else:
+                    ec2instance = False
+            except ClientError:
+                continue
+            except Exception:  # pylint: disable=W0703
+                trace = traceback.format_exc()
+                log.error(str(trace))
+                continue
+
+        return ec2instance
+
+    @staticmethod
+    def send_to_securityhub(manifest):
+        client = boto3.client('securityhub', region_name=config.get('aws', 'region'))
+        check_response = {}
+        found = False
+        try:
+            check_response = client.get_findings(Filters={'Id': [{'Value': manifest["Id"], 'Comparison': 'EQUALS'}]})
+            for _ in check_response["Findings"]:
+                found = True
+        except ClientError:
+            pass
+
+        import_response = False
+        if not found:
+            try:
+                import_response = client.batch_import_findings(Findings=[manifest])
+            except ClientError as err:
+                # Boto3 issue communicating with SH, throw the error in the log
+                log.error(str(err))
+            except Exception:  # pylint: disable=W0703
+                # Unknown error / issue, log the result
+                trace = traceback.format_exc()
+                log.error(str(trace))
+
+        return import_response
+
     def submit(self):
         log.info("Processing detection: %s", self.event.detect_description)
-        sqs_payload = self.create_payload()
-        response = self.sqs_queue.send_message(MessageBody=json.dumps(sqs_payload))
-        log.info(
-            'Detection found that meets severity threshold, sending to SQS. (SQS Message ID: %s)', str(
-                response.get("MessageId")))
+        sh_payload = self.create_payload()
+        send = False
+        if self.event.instance_id:
+            instance = self.find_instance(self.event.instance_id, self.event.device_details["mac_address"])
```

Comment:
We may need to check the `instance` for existence here.

Reply:
The instance variable may be `None`. It would be great to stop processing here and log the fact.
```
if instance is None:
    log.warning("Cannot submit detection for instance XYZ to AWS Security Hub: The instance cannot be found in EC2")
    return
```